pkg/envoy/server.go

```diff
@@ -426,17 +431,41 @@ func getPortNetworkPolicyRule(sel api.EndpointSelector, l7Parser policy.L7Parser
 }
 
 func getDirectionNetworkPolicy(l4Policy policy.L4PolicyMap, policyEnforced bool,
-	labelsMap identity.IdentityCache, deniedIdentities map[identity.NumericIdentity]bool) []*cilium.PortNetworkPolicy {
+	labelsMap identity.IdentityCache, allowedIdentities, deniedIdentities map[identity.NumericIdentity]bool) []*cilium.PortNetworkPolicy {
 	if !policyEnforced {
 		// Return an allow-all policy.
 		return allowAllPortNetworkPolicy
 	}
 
-	if len(l4Policy) == 0 {
+	if len(allowedIdentities)+len(l4Policy) == 0 {
 		return nil
 	}
 
-	PerPortPolicies := make([]*cilium.PortNetworkPolicy, 0, len(l4Policy))
+	PerPortPolicies := make([]*cilium.PortNetworkPolicy, 0, 2+len(l4Policy))
+
+	// allowedIdentities is the set of identities allowed at L3-only.
+	// Generate two allow-all L3 rules (one for TCP and one for UDP) to allow
+	// all traffic with those remote identities.
+	var l3RemotePolicies []uint64
```

Comment:
Should we have an ANY value for the protocol to avoid the duplication and also to allow other protocols?

Reply:
I thought this PR was mainly for the L3-only case (no L7), so support of arbitrary L4 should not be that hard? Have to think about making it optional. As this is not a mandatory change, maybe we file it as a "feature request" for a next version of the protocol?

Reply:
Yeah, the problem is that we're using Envoy's enum for protocol, which defaults to TCP when no value is set. We can fix that by either defining our own enum, or wrapping it into a message and making that optional.
I wouldn't worry too much about it, since we only have 2 values in this enum.

Reply:
This is using L4 expanded policy already, correct? Cilium will already have expended ANY into a separate TCP and UDP rule.

Reply:
No, this is for L3-only rules. We can't wildcard the L4 protocol, due to using Envoy's enum for the L4 protocol, which doesn't have an "any" value. Therefore, I have to expand each L3-only rule into 2 L4 rules: one for TCP, and one for UDP.
This expansion is not too bad, since we'll have either 0 or 2 such rules for any endpoint to represent the L3-only rules.

Reply:
And I agree that L4 "any" rules are already expanded at this point.